### PR TITLE
添加llm来源spark_api。

### PR DESCRIPTION
**Pull request by zhouzinimg**
由于spark_api没有官方类，所以需要额外的一个文件。

### CODE REVIEW

File: metagpt/config.py [hunk 1]
```diff
@@ -45,8 +45,18 @@ def __init__(self, yaml_file=default_yaml_file):
         self.global_proxy = self._get("GLOBAL_PROXY")
         self.openai_api_key = self._get("OPENAI_API_KEY")
         self.anthropic_api_key = self._get("Anthropic_API_KEY")
+
+        #星火大模型相关
+        self.xinghuo_appid = self._get("xinghuo_appid")
+        self.xinghuo_api_secret = self._get("xinghuo_api_secret")
+        self.xinghuo_api_key = self._get("xinghuo_api_key")
+        self.domain=self._get("domain")
+        self.Spark_url=self._get("Spark_url")
```
Comment (better629) [inline metagpt/config.py:54]: keep the format as `self.spark_url`

File: metagpt/llm.py [hunk 2]
```diff
@@ -8,9 +8,9 @@
 
 from metagpt.provider.anthropic_api import Claude2 as Claude
 from metagpt.provider.openai_api import OpenAIGPTAPI as LLM
+from metagpt.provider.spark_api import Spark
 
-DEFAULT_LLM = LLM()
-CLAUDE_LLM = Claude()
+DEFAULT_LLM = Spark()
```
Comment (better629) [inline metagpt/llm.py:13]: can't change the DEFAULT_LLM. What if the user want to use OPENAI_API. The better way is to setup a LLM which does't care about the API_CONFIG

File: metagpt/provider/SparkApi.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
+import _thread as thread
```
Comment (better629) [inline metagpt/provider/SparkApi.py:1]: format the file name as `spark_api.py` or other to keep the format.

File: metagpt/provider/SparkApi.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        "payload": {
+            "message": {
+                "text": question
+            }
+        }
+    }
+    return data
+
+
+def main(appid, api_key, api_secret, spark_url,domain, question):
```
Comment (better629) [inline metagpt/provider/SparkApi.py:127]: add param input/output type.
and the capsulation of SparkDesk python sdk seems not good enough, maybe you can have a update.

File: metagpt/provider/spark_api.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        leng = len(temp)
+        length += leng
+    return length
+
+def checklen(text):
+    while (getlength(text) > 8000):
+        del text[0]
+    return text
+
+class Spark:
```
Comment (better629) [inline metagpt/provider/spark_api.py:27]: maybe you should inherit from `BaseGPTAPI`

File: metagpt/reflect_and_retrieve/GA_memory_storage.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python
```
Comment (better629) [inline metagpt/reflect_and_retrieve/GA_memory_storage.py:1]: submit the ga_game's commits as independent pr. and it should be related the branch `ga_game` not `main`

File: metagpt/reflect_and_retrieve/Prompt_template/poignancy_chat_v1.txt [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+<commentblockmarker>###</commentblockmarker>
+Here is a brief description of !<INPUT 0>!. 
+!<INPUT 1>!
+
+On the scale of 1 to 10, where 1 is purely mundane (e.g., routine morning greetings) and 10 is extremely poignant (e.g., a conversation about breaking up, a fight), rate the likely poignancy of the following conversation for !<INPUT 2>!.
+
+Conversation: 
+!<INPUT 3>!
+
+Rate (return a number between 1 to 10):
```
Comment (better629) [inline metagpt/reflect_and_retrieve/Prompt_template/poignancy_chat_v1.txt:17]: delete `._.DS_Store` file

File: webui.py
Comment (better629) [inline webui.py:1]: remove useless file

**Comment by femto** [comment]
adds websocket to requirements.txt

**Comment by femto** [comment]
can't run?:
 File "/Users/femtozheng/python-project/MetaGPT/metagpt/roles/product_manager.py", line 40, in __init__
    self._init_actions([WritePRD])
  File "/Users/femtozheng/python-project/MetaGPT/metagpt/roles/role.py", line 112, in _init_actions
    i = action("")
  File "/Users/femtozheng/python-project/MetaGPT/metagpt/actions/write_prd.py", line 223, in __init__
    super().__init__(name, context, llm)
  File "/Users/femtozheng/python-project/MetaGPT/metagpt/actions/action.py", line 25, in __init__
    llm=LLM.DEFAULT_LLM
AttributeError: type object 'OpenAIGPTAPI' has no attribute 'DEFAULT_LLM'

File: config/config.yaml [hunk 8]
```diff
@@ -15,6 +15,19 @@ RPM: 10
...
+#### if Xinghuo
+#XINGHUO_APPID : "YOUR_APPID"
+#XINGHUO_API_SECRET : "YOUR_APISecret"
+#XINGHUO_API_KEY : "YOUR_APIKey"
+#DOMAIN : "generalv2"
+#SPARK_URL : "ws://spark-api.xf-yun.com/v2.1/chat"
+
+XINGHUO_APPID : "ae5e30f4"
+XINGHUO_API_SECRET : "MDhlOWE2NmFhOWMxZWRkOTdlYjY2Njk1"
+XINGHUO_API_KEY : "97b635fe5927d34a857333e11d15f29f"
```
Comment (better629) [inline config/config.yaml:27]: Don't commit your own key, use `YOUR_APIKEY`

File: metagpt/provider/spark_api.py [hunk 9]
```diff
@@ -5,54 +5,188 @@
...
+from metagpt.provider.base_gpt_api import BaseGPTAPI
+
+
+
+
+class SparkAPI(BaseGPTAPI):
+
+    def get_choice_text(self, rsp: dict) -> str:
+        return rsp["payload"]["choices"]["text"][-1]["content"]
+    async def acompletion_text(self, messages: list[dict], stream=False) -> str:
```
Comment (stellaHSR) [inline metagpt/provider/spark_api.py:36]: 添加换行符

File: metagpt/provider/spark_api.py [hunk 10]
```diff
@@ -5,54 +5,188 @@
...
+        #尚未实现
+        pass
+
+    async def acompletion(self, messages: list[dict]):
+        # 尚未实现
+        w = GetMessageFromWeb(messages)
+        return  await w.arun()
+
+    def completion(self, messages: list[dict]):
+        w= GetMessageFromWeb(messages)
```
Comment (stellaHSR) [inline metagpt/provider/spark_api.py:46]: 每次获取信息初始化？

File: metagpt/provider/spark_api.py [hunk 11]
```diff
@@ -5,54 +5,188 @@
...
+                "authorization": authorization,
+                "date": date,
+                "host": self.host
+            }
+            # 拼接鉴权参数，生成url
+            url = self.spark_url + '?' + urlencode(v)
+            # 此处打印出建立连接时候的url,参考本demo的时候可取消上方打印的注释，比对相同参数时生成的url与自己代码生成的url是否一致
+            return url
+
+    def __init__(self,text):
```
Comment (stellaHSR) [inline metagpt/provider/spark_api.py:100]: PEP 8

**Comment by stellaHSR** [comment]
后续可以在每次提交时的commit，写更有明确修改点的文字，便于后续自己查看和回溯。

**Comment by stellaHSR** [comment]
Good job

**Comment by stellaHSR** [comment]
LGTM

**Comment by femto** [comment]
LGTM

**Pull request merged**